src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs

```diff
@@ -26,6 +30,15 @@ public override void ValidateId(Slice lowerId)
             if (_documentDatabase.ForTestingPurposes != null && _documentDatabase.ForTestingPurposes.EnableWritesToTheWrongShard)
                 return;
 
+            if (documentFlags.Contain(DocumentFlags.FromReplication))
+            {
+                // RavenDB-21104
+                // we allow writing the document to the wrong shard to avoid inconsistent data within the shard group
+                // and handle the leftovers at the end of the transaction 
+                context.Transaction.ExecuteDocumentsMigrationBeforeCommit();
```

Comment:
shouldn't this be _after_ commit? 
what will happened if the task was completed before the actual commit?

Reply:
done in https://github.com/ravendb/ravendb/pull/17202